lib/browser/browser.ts

```diff
@@ -300,11 +321,158 @@ class OktaAuthBrowser extends OktaAuthBase implements OktaAuth, SignoutAPI {
     };
     return http.get(this, url, options);
   }
-  
+
+  //
+  // Methods from dowstream SDKs' AuthService
+  //
+
+  // Common APIs
+
+  async updateAuthState(): Promise<void> {
+    const handleUpdate = () => {
+      return Promise.all([
+        this.tokenManager.get(ACCESS_TOKEN_STORAGE_KEY),
+        this.tokenManager.get(ID_TOKEN_STORAGE_KEY)
+      ]).then(([accessToken, idToken]) => {
+        if (accessToken && this.tokenManager.hasExpired(accessToken)) {
+          accessToken = null;
+        }
+        if (idToken && this.tokenManager.hasExpired(idToken)) {
+          idToken = null;
+        }
+        return this.authStateManager._updateAuthState({ accessToken, idToken } as Tokens);
```

Comment:
I think there is a problem with this sequential approach. Your earlier iteration which was taking arguments for accessToken and idToken was dealing with this issue. In order to have sequential state updates we would need to have all parameters available **synchronously in advance**. Any parameter which is obtained asynchronously will be subject to race conditions. We are not emitting a snapshot of the current state, we are evaluating the state at a future time and emitting the result of that evaluation. So if several changes were made in quick succession (say 2 tokens were stored, and 2 updateAuthStates were queued), based on the code here - I would expect both updateAuthStates to emit the same state object. The reason is that the token is stored synchronously but is being read asynchronously. So the first token read will happen after both token writes.

So my current position is that trying to maintain sequential state updates is difficult, maybe impossible, since we cannot know all the variables that go into the `isAuthenticated` function, which will be evaluated at a moment in time. So I think the best solution is, if it is in progress, `updateAuthState` should cancel itself (do not emit) and restart. Also I think its best to not return a Promise to the caller (the caller should fire and forget)

Reply:
hmm, I still think this implementation would work. As it only put the async function into the queue (the async process, get tokens and evaluate the states, only happen in the queue), then the queue will follow a sequential manner to keep generating new states for the future. Call `oktaAuth.updateAuthState` (the one we will expose to users) is actually just like when an event is triggered in AuthStateManager.

Reply:
never mind, missed the part that write tokens to storage is not in the sequential process...